### ISSUE TITLE
TouchProcessor.enqueue is creating too many arrays

### ISSUE THREAD

**Issue by hardcoremore** [Gamua/Starling-Framework#1025]
https://github.com/Gamua/Starling-Framework/blob/365ece4199385d2e197219a2fa57114b42cb818d/starling/src/starling/events/TouchProcessor.as#L229

Can this array be pooled instead of created every time?

This method is creating too many Array instances. Take a look at the screenshot:
https://ibb.co/bSM0Zn

And here is the Scout Profile:
https://goo.gl/NYXced

This was tested on iPhone 6s.

**Comment by gevorg-gsar** [comment]
@hardcoremore good catch! 
@PrimaryFeather any ideas? Can this be also related to Android ANR-s related to touch?

**Comment by hardcoremore** [comment]
Now that I think about it, I think it is a very easy solution to this. It is a little more lines of code but far less allocations. I think this can be safely replaced with this:

_queue.unshift(_touchMarker.mockY);
_queue.unshift(_touchMarker.mockX);
_queue.unshift(phase);
_queue.unshift(1);


What do you guys think?

**Comment by JohnBlackburne** [comment]
That code only runs when using multitouch simulation, so it should not show up when profiling a production build, either on mobile or desktop.

**Comment by hardcoremore** [comment]
Yeah, it is strange. Well I need to test that again with Scout to see if it is really running when multi touch simulation is not used.

**Comment by PrimaryFeather** [comment]
Sorry for the late reply, I somehow overlooked this post!

In any case, that code only runs for multitouch simulation, so that shouldn't be an issue. Furthermore, you cannot replace it with multiple `unshift` calls, because the queue needs to be filled with _Arrays_ of arguments, not the arguments themselves.

And even if we could, it would have the contrary effect: since `unshift` uses variable arguments (`... args`), it will create one array per call (!) on iOS. (This is the same reason why `Array.push()` is only used very rarely in Starling.)

In other words, I'd rather leave this as it is. :wink:

**Comment by hardcoremore** [comment]
Hi Daniel,

Thanks for the replay. Can you just please give a little more details about:

"it will create one array per call (!) on iOS. (This is the same reason why Array.push() is only used very rarely in Starling."

Is that a new bug for iOS I am not aware? Why is Array.push() avoided in Starling?

**Comment by PrimaryFeather** [comment]
> "it will create one array per call (!) on iOS. (This is the same reason why Array.push() is only used very rarely in Starling.)"
> Is that a new bug for iOS I am not aware? Why is Array.push() avoided in Starling?

No, this is not new; it's just something the AOT compiler does not optimize (or cannot, for whatever reason). Whenever you've got a method with a variable number of arguments, an array will be created and filled with the parameters you passed into the method. The JIT-compiler seems to be able to optimize this, but the AOT compiler can't.

_Array.push()_ is another method with variable arguments, so it's better avoided, too. _Array.insertAt()_ can be used instead, though — that was added on Josh's and my recommendation as a workaround.

**Comment by hardcoremore** [comment]
Oh so whenever I use Array.push or Vector.push an array will be allocated regardless if I just pass a object as single parameter?

**Comment by PrimaryFeather** [comment]
Exactly! On iOS, at least.

**Comment by hardcoremore** [comment]
Ok great thanks. I already started to refactor to eliminate all push. Sigh, so many little different things for each platform :)

**Comment by PrimaryFeather** [comment]
Yeah, I can feel with you. :wink:
Very annoying indeed!

**Comment by hardcoremore** [comment]
I can confirm 100% that TouchProcessor.enqueue is creating too much arrays. 

This is because of using unshift method. It is affecting Android and iOS as well.

Take a look at the Adobe Scout profile screenshot:

https://ibb.co/bTws6T


I have refactored this to not use unshift method.

**Comment by JohnBlackburne** [comment]
It looks like you are profiling a desktop version again, not Android or iOS. Apart from unshift only being called if _ctrlDown is set, which can only happen on desktop, your screenshot says Scout recorded processing of a mouseMove event which can only happen on desktop.

**Comment by hardcoremore** [comment]
Hi John,

No I am not profiling Desktop version. Those allocations are happening not just because of unsifht but because using arguments parameter as well.

**Comment by PrimaryFeather** [comment]
Hm, yeah, it could be that using `arguments` is another way to force an array creation. I'll look into it!

**Comment by hardcoremore** [comment]
Thanks Daniel.

This is my version of TouchProcessor class that I now using. I also implemented poling of Touch object as well:

https://pastebin.com/RfAXb1MG

And here is the TouchData class:

https://pastebin.com/9xdXfbzv


Thanks

**Comment by PrimaryFeather** [comment]
Thanks for sharing the code!

One problem that I noticed is that this change will break code of developers who extended the TouchProcessor class, since the `_queue` member is protected. This will probably not have been done very often, but still ... the question remains if this optimization is worth breaking existing code.

**Comment by hardcoremore** [comment]
Hi Daniel,

Yes, unfortunately it breaks backwards compatibility. I don't know if that change is worth for general uses but it is definitely worth for me.  I am now running Starling from my branch.

My guess is that it is probably worth it for general use case as well because it prevents allocations which is always good and change is not that big and probably very little percent of people are extending that class.

**Comment by PrimaryFeather** [comment]
I tend to agree, but I'll take another look! Cheers!

**Comment by PrimaryFeather** [comment]
I finally committed this! I agree that the compatibility issue is a price worth paying in this case.

**Comment by hardcoremore** [comment]
Awesome, thanks!
